pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -391,173 +409,172 @@ func validateParamArrayIndex(ctx context.Context, params []v1beta1.Param, spec *
 			if p.Default.Type == v1beta1.ParamTypeArray {
 				for _, pattern := range patterns {
 					for i := 0; i < len(p.Default.ArrayVal); i++ {
-						arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
+						arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
 					}
 				}
 			}
 		}
 	}
 
-	// Collect array params lengths from pipeline
+	// Collect array params lengths from params
 	for _, p := range params {
 		if p.Value.Type == v1beta1.ParamTypeArray {
 			for _, pattern := range patterns {
 				for i := 0; i < len(p.Value.ArrayVal); i++ {
-					arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
+					arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
 				}
 			}
 		}
 	}
+	return arrayParamsLengths
+}
 
+// ValidateOutofBoundArrayParams validates if the array indexing params are out of bound
+func ValidateOutofBoundArrayParams(arrayIndexingParams []string, arrayParams map[string]int) error {
```

Comment:
I'm nitpicking but changing `arrayParams` to `arrayParamsLengths` can be clearer.